clang/lib/CConv/ConstraintVariables.cpp

```diff
@@ -1647,19 +1645,57 @@ void FunctionVariableConstraint::mergeDeclaration(ConstraintVariable *FromCV) {
   auto RetVar = getOnly(returnVars);
   RetVar->mergeDeclaration(FromRetVar);
   // Transplant params.
-  assert(From->numParams() == numParams());
-  for (unsigned i = 0; i < From->numParams(); i++) {
-    CVarSet &FromP = From->getParamVar(i);
-    CVarSet &P = getParamVar(i);
-    auto FromVar = getOnly(FromP);
-    auto Var = getOnly(P);
-    Var->mergeDeclaration(FromVar);
+  handle_params(From, this,
+                [](ConstraintVariable *F, ConstraintVariable *T)
+                { T->mergeDeclaration(F); });
+}
+
+void FunctionVariableConstraint::handle_params
+(FVConstraint *From, FVConstraint *To,
+ std::function<void(ConstraintVariable*,ConstraintVariable*)> f) {
+  bool fromEmpty = From->numParams() == 0;
+  bool toEmpty = To->numParams() == 0;
+  bool paramsEq = From->numParams() == To->numParams();
+  assert(fromEmpty || toEmpty || paramsEq);
+  if (paramsEq) {
+    for (unsigned i = 0; i < From->numParams(); i++) {
+      CVarSet &FromP = From->getParamVar(i);
+      CVarSet &P = To->getParamVar(i);
+      auto FromVar = getOnly(FromP);
+      auto Var = getOnly(P);
+      f(FromVar, Var);
+    }
+  } else { // Dealing with an untyped prototype
+    FVConstraint *Empty = fromEmpty ? From : To;
+    FVConstraint *Typed = fromEmpty ? To : From;
+    for (auto deferred : Empty->getDeferredParams()) {
+      assert(Typed->numParams() == deferred.PS.size());
+      for(unsigned i = 0; i < deferred.PS.size(); i++) {
+        CVarSet ParamDC = Typed->getParamVar(i);
+        CVarSet ArgDC = deferred.PS[i];
+        ProgramInfo &Info = deferred.I;
+        auto &CS = Info.getConstraints();
+        constrainConsVarGeq(ParamDC, ArgDC, CS, &(deferred.PL), Wild_to_Safe, false, &Info);
+      }
+    }
```

Comment:
You still need to merge the `Var`s. When you had two prototypes like `f(int *x)` and `f(int *)` and were merging/brain transplanting, you update the internal `cvars` of `ConstraintVariable`(s) in question. Here, if you are brain-transplanting a `f()` style prototype with `f(int *x, int *y)` then you need to move the latter's internal `cvars` into the former. Before we assumed the arity of the from/to was the same, and then the `handle_params` code merges them. But now that the arity is different, you don't merge, you just move.